internal/function/diff.go

```diff
@@ -0,0 +1,292 @@
+
+type (
+	// Diff implements git-diff function as a UDF
+	Diff struct {
+		repo sql.Expression
+		from sql.Expression
+		to   sql.Expression
+	}
+
+	// diffItem stands for a single payload item
+	// (response payload is an array of JSON objects as strings)
+	// Indexes are local per file.
+	// Many objects for different files may have the same index.
+	diffItem struct {
+		Index  int    `json:"index"`
+		File   string `json:"file"`
+		Chunk  string `json:"chunk"`
+		Status string `json:"status"`
+	}
+)
+
+// NewDiff constructor
+func NewDiff(args ...sql.Expression) (sql.Expression, error) {
+	d := &Diff{}
+	switch len(args) {
+	case 2:
+		d.repo, d.to = args[0], args[1]
+	case 3:
+		d.repo, d.from, d.to = args[0], args[1], args[2]
+	default:
+		return nil, sql.ErrInvalidArgumentNumber.New("DIFF", "2 or 3", len(args))
+	}
+
+	return d, nil
+}
+
+func (d *Diff) String() string {
+	if d.from == nil {
+		return fmt.Sprintf("diff(%s, %s)", d.repo, d.to)
+	}
+
+	return fmt.Sprintf("diff(%s, %s, %s)", d.repo, d.from, d.to)
+}
+
+// Type implements the sql.Expression interface.
+func (*Diff) Type() sql.Type {
+	return sql.Array(sql.JSON)
+}
+
+// TransformUp implements the Expression interface.
+func (d *Diff) TransformUp(fn sql.TransformExprFunc) (sql.Expression, error) {
+	repo, err := d.repo.TransformUp(fn)
+	if err != nil {
+		return nil, err
+	}
+
+	to, err := d.to.TransformUp(fn)
+	if err != nil {
+		return nil, err
+	}
+
+	if d.from == nil {
+		return fn(&Diff{repo: repo, to: to})
+	}
+
+	from, err := d.from.TransformUp(fn)
+	if err != nil {
+		return nil, err
+	}
+	return fn(&Diff{repo: repo, from: from, to: to})
+}
+
+// Children implements the Expression interface.
+func (d *Diff) Children() []sql.Expression {
+	if d.from == nil {
+		return []sql.Expression{d.repo, d.to}
+	}
+	return []sql.Expression{d.repo, d.from, d.to}
+}
+
+// IsNullable implements the Expression interface.
+func (*Diff) IsNullable() bool {
+	return false
+}
+
+// Resolved implements the Expression interface.
+func (d *Diff) Resolved() bool {
+	return d.to.Resolved() && d.to.Resolved() && (d.from == nil || d.from.Resolved())
+}
+
+// Eval implements the sql.Expression interface.
+func (d *Diff) Eval(ctx *sql.Context, row sql.Row) (interface{}, error) {
+	span, ctx := ctx.Span("gitbase.Diff")
+	defer span.Finish()
+
+	repo, err := d.getRepo(ctx, row)
+	if err != nil {
+		return nil, err
+	}
+
+	from, to, err := d.getCommits(ctx, repo, row)
+
+	return diff(ctx, from, to)
+}
+
+func (d *Diff) getRepo(ctx *sql.Context, r sql.Row) (*gitbase.Repository, error) {
+	repoID, err := exprToString(ctx, d.repo, r)
+	if err != nil {
+		return nil, err
+	}
+	s, ok := ctx.Session.(*gitbase.Session)
+	if !ok {
+		return nil, gitbase.ErrInvalidGitbaseSession.New(ctx.Session)
+	}
+	return s.Pool.GetRepo(repoID)
+}
+
+func (d *Diff) getCommits(ctx *sql.Context, repo *gitbase.Repository, row sql.Row) (*object.Commit, *object.Commit, error) {
+	commitHash, err := exprToString(ctx, d.to, row)
```

Comment:
The function should accept revisions and not only hashes, take a look at: [Repository.ResolveRevision](https://godoc.org/gopkg.in/src-d/go-git.v4#Repository.ResolveRevision).

This can be skipped in the first version if the usage of this is not straight forward.

Reply:
Added `ResolveRevision`, so first we'll try to resolve revision, if not the input string will be treated as a commit hash. But at the end the diff works with commit objects.